docs/proposals/petset.md

```diff
@@ -0,0 +1,533 @@
+most amenable to automatic generation.
+
+Because that pod needs access to stable storage, the PetSet may specify a template for one or more
+**supporting resources** that can be used for each distinct pod. In the future other resources may be
+added that must also be templated - for instance, secrets (unique secret per member), config data (unique
+config per member), and in the future, arbitrary extension resources.
+
+The set of parameterization allowed to each pod and supporting resources is intentionally limited to those
+that accomplish the goals above. It should be possible to:
+
+* Identify the supporting resources by name in the pod template (pvc name)
+* Give the supporting resource a predictable name based on identity
+* Set environment variables that correspond to the identity of the pod for use by the software
+* Provide the current desired set size (`spec.replicas`) to each pod as an integer
+
+```
+type PetSetSpec struct {
+    ...
+    // The template to use for the pod in the set
+    Template *PodTemplateSpec
+    SupportTemplates []SupportTemplateSpec
+}
+
+type SupportTemplateSpec struct {
+    // only one of these fields may be set, like Volumes
+    PersistentVolumeClaim *PersistentVolumeClaimTemplateSpec
+    Secret *SecretTemplateSpec
+    ConfigData *ConfigDataTemplateSpec
+    // anything not supported by the above
+    Object runtime.Object
+}
+
+// Proposal 2a: use the name of the support resource as identity implicitly
+// Downside: Requires the controller to understand the places to be parameterized in the template - prevents
+//   custom resources from being supported.
+{
+    "spec": {
+        "template": {
+            "spec": {
+                "volumes": [{
+                    "persistentVolumeClaim": {
+                        "claimName": "pod-name-pvc" // controller appends IDENTITY to this claim automatically
+                    }
+                }]
+            }
+        },
+        "supportTemplates": [{
+            "persistentVolumeClaim": {
+                "metadata": {"name": "pod-name-pvc"} // controller appends IDENTTY to this claim automatically
+            }
+        }]
+    }
+}
+
+// Proposal 2b: append identity to specified field paths
+// Downside: Flexibility is limited, may need "operations" on the field in order to transform value
+{
+    "spec": {
+        // ...
+
+        "supportTemplates": [{
+            "persistentVolumeClaim": {
+                "metadata": {"name": "pod-name-pvc-"}
+            },
+            "fieldPaths": ["metadata.name"] // controller appends IDENTTY to field metadata.name
+        }]
+    }
+}
+
+// Proposal 2c: append identity to specified field paths, name is special
+// Downside: Flexibility is limited, transformations may be required
+{
+    "spec": {
+        // ...
+
+        "supportTemplates": [{
+            "configData": {
+                "metadata": {"generateName": "pod-name-pvc-"}
```

Comment:
Though if we appended identity automatically to references for implicit scoping, I'd just use name here to match.